packages/core/nano/src/types.ts

```diff
@@ -1,22 +1,139 @@
-export interface Consumer<T> {
-    notify: <K extends keyof T>(property: K, value: T[K]) => void;
-    notifyMany: <K extends keyof T>(values: Pick<T, K>) => void;
+export interface Pending<T> {
+  kind: "Pending";
+  estimate?: T;
 }
 
-export interface Pending<T> {
-    kind: "Pending";
-    estimate?: T;
+/**
+ * The interface for an object whose data can be bound to. We use this contract for
+ * components that want to expose their data and its changes to other components.
+ * An example of this would be Formula binding to a Tablero. In this scenario,
+ * Tablero is expected to implement IProducer and expose some of its data, given
+ * a reference, through getDataResult. Formula is expected to implement IConsumer
+ * and register itself as an observer to Tablero.
+ *
+ * Any component that implements IProducer is expected to provide some registration
+ * functionality and to notify observers whenever the data they are bound to changes.
+ */
+export interface IProducer<T> {
+  /**
+   * Binds the given observer to the component associated with this Producer.
+   * @param consumer - The object to start listening to any changes in the bound data.
+   */
+  addConsumer(consumer: IConsumer<T>): void;
+
+  /**
+   * Unbinds the given observer from the component associated with this Producer.
+   * @param consumer - The observer to unregister from the Producer.
+   */
+  removeConsumer(consumer: IConsumer<T>): void;
+
+  /**
+   * Return IDataResult for the given dataBindingReference, containing the data value
+   * and some additional metadata.
+   * @param dataBindingReference - An instance of dataBindingReference, a reference
+   * to some data in the bound component.
+   */
+  read<K extends keyof T>(consumer: IConsumer<T>, property: K, ...args: any[]): T[K] | Pending<T[K]>;
+
+  /**
+   * Invoked when this data provider is deleted.
+   */
+  onDeleted(): void;
+}
+
+/**
+ * The interface for an object that can bind to an IProducer.
+ * One example of this provider-observer relationship is a Formula binding
+ * to Tablero. In this scenario, Formula is an IConsumer listening to changes
+ * in the IProducer, Tablero.
+ *
+ * Any object that implements IConsumer is expected to provide a
+ * callback whenever the component it is bound to changes in value and a reference
+ * to the data that the observer is bound to.
+ */
+export interface IConsumer<T> {
+  /**
+   * Invoked whenever the data this object is bound to is changed.
+   */
+  notify: <K extends keyof T>(property: K, value: T[K]) => void;
```

Comment:
If a consumer can subscribe to multiple producers then I think the paramater would be needed in notify. In this case (and perhaps regardless), I wonder if producers should have a string id on them to make them easy to identify (rather than relying on object identity).